From: Anthony Romano <anthony.romano@coreos.com>
Subject: [PATCH] tmpfs: don't undo fallocate past its last page
Date: Sun,  8 May 2016 06:16:27 -0700
Cc: linux-mm@kvack.org, linux-kernel@vger.kernel.org, Anthony Romano
	<anthony.romano@coreos.com>
To: hughd@google.com
Archived-At: <http://permalink.gmane.org/gmane.linux.kernel.mm/151142>

When fallocate is interrupted it will undo a range that extends one byte
past its range of allocated pages. This can corrupt an in-use page by
zeroing out its first byte. Instead, undo using the inclusive byte range.

[HH: backported to 4.4.x]
Signed-off-by: Anthony Romano <anthony.romano@coreos.com>
---
 mm/shmem.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mm/shmem.c b/mm/shmem.c
index 719bd6b..f0f9405 100644
--- a/mm/shmem.c
+++ b/mm/shmem.c
@@ -2155,7 +2155,7 @@ static long shmem_fallocate(struct file *file, int mode, loff_t offset,
 			/* Remove the !PageUptodate pages we added */
 			shmem_undo_range(inode,
 				(loff_t)start << PAGE_CACHE_SHIFT,
-				(loff_t)index << PAGE_CACHE_SHIFT, true);
+				((loff_t)index << PAGE_CACHE_SHIFT) - 1, true);
 			goto undone;
 		}
 
-- 
2.8.1
